pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -516,15 +515,12 @@ func (c *Controller) Network(endpointIP string, labels labels.Instance) network.
 }
 
 func (c *Controller) Cleanup() error {
-	svcs, err := c.serviceLister.List(klabels.Everything())
-	if err != nil {
-		return fmt.Errorf("error listing services for deletion: %v", err)
+	if err := queue.WaitForClose(c.queue, 30*time.Second); err != nil {
```

Comment:
What is this for?

Reply:
We _must_ make sure the queue is actually done. If we don't we could end up removing endpointshards, then processing an event _after_ that, and the endpoint for the removed cluster will linger forever.

Reply:
Got it, suddenly i think of the below closing of  `stop` better move in front of `handleDelete`

```
func (c *Controller) deleteCluster(secretKey string, clusterID cluster.ID) {
	c.cs.Lock()
	defer func() {
		c.cs.Unlock()
		log.Infof("Number of remote clusters: %d", c.cs.Len())
	}()
	log.Infof("Deleting cluster_id=%v configured by secret=%v", clusterID, secretKey)
	err := c.handleDelete(clusterID)
	if err != nil {
		log.Errorf("Error removing cluster_id=%v configured by secret=%v: %v",
			clusterID, secretKey, err)
	}
	close(c.cs.remoteClusters[secretKey][clusterID].stop)
	delete(c.cs.remoteClusters[secretKey], clusterID)
}
```

Reply:
moved